File: langs/typescript/tests/visual/snapshot-all-tabs.mjs
/**
 * snapshot-all-tabs.mjs
 *
 * Visual snapshot script for GuideArch TypeScript UI.
 * Runs headless Chromium, captures all tabs in empty and SAS-loaded states,
 * then tests that the weight edit updates the Results chart.
 *
 * Usage: node tests/visual/snapshot-all-tabs.mjs
 *
 * Prerequisites:
 *   1. pnpm dev must be startable (runs in background here)
 *   2. Playwright chromium must be installed:
 *      pnpm exec playwright install chromium
 */

import { chromium } from 'playwright';
import { spawn } from 'child_process';
import { fileURLToPath } from 'url';
import path from 'path';
import fs from 'fs';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const SNAPSHOTS_DIR = path.join(__dirname, 'snapshots');
const TS_DIR = path.join(__dirname, '..', '..');

// Ensure snapshots dir exists
fs.mkdirSync(SNAPSHOTS_DIR, { recursive: true });

const APP_URL = 'http://localhost:1420/';
const VIEWPORT = { width: 1440, height: 900 };

const TABS = [
  'Decisions',
  'Alternatives',
  'Properties',
  'Coefficients',
  'Constraints',
  'Results',
  'Critical Decisions',
  'Critical Constraints',
];

// ---------------------------------------------------------------------------
// Dev server management
// ---------------------------------------------------------------------------

async function startDevServer() {
  console.log('[dev] Starting pnpm dev...');
  const proc = spawn('pnpm', ['dev'], {
    cwd: TS_DIR,
    stdio: ['ignore', 'pipe', 'pipe'],
    detached: false,
  });

  proc.stderr.on('data', (d) => {
    const msg = d.toString().trim();
    if (msg) console.log('[dev err]', msg);
  });

  // Wait for the server to report it's ready
  await new Promise((resolve, reject) => {
    const timeout = setTimeout(() => reject(new Error('Dev server timed out after 30s')), 30000);
    proc.stdout.on('data', (d) => {
      const text = d.toString();
      if (text.includes('localhost') || text.includes('ready') || text.includes('Local:')) {
        clearTimeout(timeout);
        resolve(undefined);
      }
    });
    proc.on('error', (err) => {
      clearTimeout(timeout);
      reject(err);
    });
    proc.on('exit', (code) => {
      if (code !== null) {
        clearTimeout(timeout);
        reject(new Error(`Dev server exited with code ${code}`));
      }
    });
  });

  console.log('[dev] Server started.');
  return proc;
}

async function waitForPort(port, timeoutMs = 15000) {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    try {
      const resp = await fetch(`http://localhost:${port}/`);
      if (resp.ok || resp.status === 200) return;
    } catch {
      // not ready yet
    }
    await new Promise((r) => setTimeout(r, 300));
  }
  throw new Error(`Port ${port} not ready after ${timeoutMs}ms`);
}

// ---------------------------------------------------------------------------
// Screenshot helpers
// ---------------------------------------------------------------------------

async function clickTab(page, tabName) {
  // Use exact text match to avoid "Decisions" matching "Critical Decisions"
  const btn = page.locator('nav.tab-strip button').filter({ hasText: new RegExp(`^${tabName}$`) });
  await btn.click();
  // Small settle for reactive updates
  await page.waitForTimeout(200);
}

async function screenshot(page, filename) {
  const filepath = path.join(SNAPSHOTS_DIR, filename);
  await page.screenshot({ path: filepath, fullPage: false });
  const stat = fs.statSync(filepath);
  const kb = (stat.size / 1024).toFixed(0);
  console.log(`  [snap] ${filename} — ${kb} KB`);
  return filepath;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  let devServer = null;

  try {
    // 1. Start dev server
    devServer = await startDevServer();
    await waitForPort(1420);
    console.log('[ready] App reachable at', APP_URL);

    // 2. Launch Chromium
    const browser = await chromium.launch({ headless: true });
    const context = await browser.newContext({
      viewport: VIEWPORT,
      colorScheme: 'dark',
    });
    const page = await context.newPage();

    // 3. Load the app
    await page.goto(APP_URL, { waitUntil: 'networkidle' });
    await page.waitForTimeout(500);

    // ── Empty state screenshots ──────────────────────────────────────────────
    console.log('\n[step 4] Capturing empty state of all tabs...');
    for (const tab of TABS) {
      await clickTab(page, tab);
      const slug = tab.toLowerCase().replace(/ /g, '-');
      await screenshot(page, `empty-${slug}.png`);
    }

    // ── Load SAS sample ──────────────────────────────────────────────────────
    console.log('\n[step 5] Opening Sample SAS...');
    // Click the "Open Sample SAS" button in the toolbar
    await page.click('button.btn-sample:has-text("Open Sample SAS")');
    // Wait for candidates to appear — the Results tab should show data
    await page.waitForTimeout(1000);

    // ── SAS-loaded screenshots ───────────────────────────────────────────────
    console.log('\n[step 6] Capturing SAS-loaded state of all tabs...');
    for (const tab of TABS) {
      await clickTab(page, tab);
      const slug = tab.toLowerCase().replace(/ /g, '-');
      await screenshot(page, `sas-${slug}.png`);
    }

    // ── Edit weight and re-screenshot results ────────────────────────────────
    console.log('\n[step 7] Editing a property weight and capturing Results...');
    // Go to Properties tab
    await clickTab(page, 'Properties');
    await page.waitForTimeout(300);

    // Find the first weight input and change its value
    const weightInput = page.locator('.weight-input').first();
    (await weightInput.triple_click)
      ? weightInput.triple_click()
      : weightInput.click({ clickCount: 3 });
    await weightInput.fill('5');
    await weightInput.press('Enter');
    await page.waitForTimeout(500);

    // Switch to Results tab and screenshot
    await clickTab(page, 'Results');
    await page.waitForTimeout(500);
    await screenshot(page, 'sas-after-edit-results.png');

    // ── Done ─────────────────────────────────────────────────────────────────
    await browser.close();

    console.log('\n[done] All snapshots saved to:', SNAPSHOTS_DIR);
    const files = fs.readdirSync(SNAPSHOTS_DIR).filter((f) => f.endsWith('.png'));
    console.log(`       ${files.length} files:`, files.join(', '));
  } finally {
    // 8. Kill dev server
    if (devServer) {
      console.log('\n[dev] Killing dev server...');
      devServer.kill('SIGTERM');
      // Give it a moment to clean up
      await new Promise((r) => setTimeout(r, 500));
    }
  }
}

main().catch((err) => {
  console.error('[error]', err);
  process.exit(1);
});

File: langs/python/tests/visual/snapshot_all_tabs.py
"""Visual snapshot script — captures screenshots of every UI tab.

Usage:
    cd langs/python && uv run python tests/visual/snapshot_all_tabs.py

What it does:
  1. Starts `uv run guidearch --port 8795` in the background
  2. Launches Chromium at 1440x900 viewport, dark color scheme
  3. Loads http://localhost:8795/
  4. Screenshots empty state of every tab → snapshots/empty-<tab>.png
  5. Clicks "Open Sample SAS"
  6. Screenshots loaded state of every tab → snapshots/sas-<tab>.png
  7. Edits a property weight, screenshots Results tab again → snapshots/sas-after-edit-results.png
  8. Kills the server

Tabs covered: Decisions, Alternatives, Properties, Coefficients, Constraints,
Results, Critical Decisions, Critical Constraints
"""

from __future__ import annotations

import subprocess
import sys
import time
from pathlib import Path

from playwright.sync_api import Page, ViewportSize, sync_playwright

# ---------------------------------------------------------------------------
# Config
# ---------------------------------------------------------------------------

_PORT = 8795
_BASE_URL = f"http://localhost:{_PORT}"
_SNAPSHOTS_DIR = Path(__file__).parent / "snapshots"
_VIEWPORT: ViewportSize = {"width": 1440, "height": 900}

# Tab names as they appear in the UI
_TAB_NAMES = [
    "Decisions",
    "Alternatives",
    "Properties",
    "Coefficients",
    "Constraints",
    "Results",
    "Critical Decisions",
    "Critical Constraints",
]

# Slug used for filenames
_TAB_SLUGS = [
    "decisions",
    "alternatives",
    "properties",
    "coefficients",
    "constraints",
    "results",
    "critical-decisions",
    "critical-constraints",
]


def _slug(name: str) -> str:
    return name.lower().replace(" ", "-")


def _wait_for_server(url: str, timeout: float = 30.0) -> bool:
    """Poll until the server responds or timeout."""
    import urllib.request

    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            urllib.request.urlopen(url, timeout=2)
            return True
        except Exception:
            time.sleep(0.5)
    return False


def _screenshot_all_tabs(page: Page, prefix: str) -> list[str]:
    """Click each tab and take a screenshot. Returns list of saved paths."""
    saved: list[str] = []
    for name, slug in zip(_TAB_NAMES, _TAB_SLUGS, strict=False):
        # Click the tab — use exact=True to avoid strict mode violations for
        # tabs like "Decisions" that also match "Critical Decisions"
        try:
            page.get_by_role("tab", name=name, exact=True).click()
            page.wait_for_timeout(900)  # let content render
        except Exception as exc:
            print(f"  Warning: could not click tab '{name}': {exc}")
            # Fallback: locate by tab text content
            try:
                page.locator(f".q-tab:has(.q-tab__label:text-is('{name}'))").click()
                page.wait_for_timeout(900)
            except Exception as exc2:
                print(f"  Fallback also failed: {exc2}")
                continue

        out_path = _SNAPSHOTS_DIR / f"{prefix}-{slug}.png"
        page.screenshot(path=str(out_path), full_page=False)
        size = out_path.stat().st_size if out_path.exists() else 0
        print(f"  Saved: {out_path.name}  ({size // 1024} KB)")
        saved.append(str(out_path))
    return saved


def main() -> None:
    _SNAPSHOTS_DIR.mkdir(parents=True, exist_ok=True)

    # ── 1. Start server ───────────────────────────────────────────────────────
    print(f"Starting guidearch on port {_PORT}...")
    # We need to run from the langs/python directory using uv
    _python_dir = Path(__file__).parents[2]
    import shutil

    uv_bin = shutil.which("uv") or "uv"
    server_proc = subprocess.Popen(
        [uv_bin, "run", "guidearch", "--port", str(_PORT)],
        cwd=str(_python_dir),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    print(f"Waiting for server at {_BASE_URL}...")
    alive = _wait_for_server(_BASE_URL, timeout=30)
    if not alive:
        server_proc.kill()
        print("ERROR: Server did not start within 30 seconds.")
        sys.exit(1)

    print("Server is up.")
    time.sleep(2)  # extra settle time for NiceGUI

    try:
        with sync_playwright() as pw:
            browser = pw.chromium.launch(
                headless=True,
                args=["--no-sandbox", "--disable-dev-shm-usage"],
            )
            ctx = browser.new_context(
                viewport=_VIEWPORT,
                color_scheme="dark",
                device_scale_factor=1,
            )
            page = ctx.new_page()

            # ── 2. Load the app ───────────────────────────────────────────────
            # Use a dedicated page for empty-state screenshots to avoid VM state
            # pollution from any previous run.
            print(f"Loading {_BASE_URL}...")
            page.goto(_BASE_URL, wait_until="networkidle", timeout=30000)
            page.wait_for_timeout(3000)  # extra settle for NiceGUI JS hydration

            # ── 3. Screenshot empty state of every tab ────────────────────────
            # Capture empty state FIRST before loading any scenario.
            # The VM starts with no scenario on a fresh server start.
            # If decisions appear (from a prior session), click "New" to clear.
            status_bar = page.locator(".guidearch-statusbar")
            if status_bar.count() > 0:
                status_text = status_bar.inner_text()
                if "No scenario" not in status_text and status_text.strip():
                    state_preview = status_text[:40]
                    print(
                        f"  Server has existing scenario state ({state_preview}), clicking New..."
                    )
                    try:
                        page.get_by_role("button", name="New").click()
                        page.wait_for_timeout(1500)
                    except Exception as exc:
                        print(f"  Warning: could not click New: {exc}")

            print("\n=== Empty state screenshots ===")
            _screenshot_all_tabs(page, "empty")

            # ── 4. Click "Open Sample SAS" ────────────────────────────────────
            print("\nClicking 'Open Sample SAS'...")
            try:
                # The button text from main.py is "Open Sample SAS"
                btn = page.get_by_role("button", name="Open Sample SAS")
                btn.click()
                page.wait_for_timeout(3000)  # wait for solve
                print("  SAS loaded.")
            except Exception as exc:
                print(f"  Warning: could not click 'Open Sample SAS': {exc}")
                # Try alternative — look for any button with SAS in it
                try:
                    page.locator("button:has-text('SAS')").first.click()
                    page.wait_for_timeout(3000)
                    print("  SAS loaded (fallback).")
                except Exception as exc2:
                    print(f"  Fallback also failed: {exc2}")

            # ── 5. Screenshot loaded state of every tab ───────────────────────
            print("\n=== SAS loaded state screenshots ===")
            _screenshot_all_tabs(page, "sas")

            # ── 6. Edit a property weight and re-screenshot Results ───────────
            print("\nEditing property weight...")
            try:
                # Go to Properties tab
                page.get_by_role("tab", name="Properties", exact=True).click()
                page.wait_for_timeout(800)

                # Find the first weight input (a number field with label "Weight")
                weight_inputs = page.locator("input[type='number']")
                cnt = weight_inputs.count()
                if cnt > 0:
                    first_weight = weight_inputs.first
                    first_weight.click()
                    first_weight.fill("5")
                    # Trigger blur by pressing Tab
                    first_weight.press("Tab")
                    page.wait_for_timeout(2000)
                    print(f"  Weight edited ({cnt} number inputs found).")
                else:
                    print("  No number inputs found; skipping weight edit.")
            except Exception as exc:
                print(f"  Weight edit failed: {exc}")

            # Screenshot Results tab after edit
            try:
                page.get_by_role("tab", name="Results").click()
                page.wait_for_timeout(1500)
                out_path = _SNAPSHOTS_DIR / "sas-after-edit-results.png"
                page.screenshot(path=str(out_path), full_page=False)
                size = out_path.stat().st_size if out_path.exists() else 0
                print(f"\nSaved: {out_path.name}  ({size // 1024} KB)")
            except Exception as exc:
                print(f"  Could not screenshot Results after edit: {exc}")

            ctx.close()
            browser.close()

    finally:
        # ── 7. Kill the server ────────────────────────────────────────────────
        print("\nKilling server...")
        server_proc.terminate()
        try:
            server_proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            server_proc.kill()
        print("Done.")


if __name__ == "__main__":
    main()
